README.md
# Engineering Challenge

## Check Resutls
  1.  To check the problem 1 open the core.clj file. You will see tax-filter-19, ret-filter-1 and Sym-diff-tax-ret functions that returns specific items inside the invoice that meet the requirements of the query. 
  2. To check the problem 2 open invoice_spec.clj. You will see the function encharged of parse the Json invoice to the new format with accurate dates and structure.
  3. To check the problem 3 open invoice_item.clj. You will see all the edge test that checks the subtotal function of the invoices. It makes it using testitems defined during the tests. 
  
## Getting started

This clojure challenge is made up of 3 questions that reflect the learning you accumulated for the past week. Complete the following instructions:

1. Create a Gitlab repo of your own and fork this repo. When complete, send us the link to your challenge results.
2. Duration: About 4 hours
3. Install Cursive Plugin to Intellij and setup a clojure deps project. https://cursive-ide.com/userguide/deps.html  

1. Enjoy!

## Problems
### Problem 1 Thread-last Operator ->>
Given the invoice defined in **invoice.edn** in this repo, use the thread-last ->> operator to find all invoice items that satisfy the given conditions. Please write a function that receives an invoice as an argument and returns all items that satisfy the conditions described below.
#### Requirements
- Load invoice to play around with the function like this:

```
(def invoice (clojure.edn/read-string (slurp "invoice.edn")))
```

#### Definitions
- An invoice item is a clojure map { … } which has an :invoice-item/id field. EG.

```
{:invoice-item/id     "ii2"  
  :invoice-item/sku "SKU 2"}
```

- An invoice has two fields :invoice/id (its identifier) and :invoice/items a vector of invoice items

#### Invoice Item Conditions
- At least have one item that has :iva 19%
- At least one item has retention :ret\_fuente 1%
- Every item must satisfy EXACTLY one of the above two conditions. This means that an item cannot have BOTH :iva 19% and retention :ret\_fuente 1%.
## Problem 2: Core Generating Functions
  Given the invoice defined in **invoice.json** found in this repo, generate an invoice that passes the spec **::invoice** defined in **invoice-spec.clj**. Write a function that as an argument receives a file name (a JSON file name in this case) and returns a clojure map such that

```
(s/valid? ::invoice invoice) => true 
```

where invoice represents an invoice constructed from the JSON.
## Problem 3: Test Driven Development
Given the function **subtotal** defined in **invoice-item.clj** in this repo, write at least five tests using clojure core **deftest** that demonstrates its correctness. This subtotal function calculates the subtotal of an invoice-item taking a discount-rate into account. Make sure the tests cover as many edge cases as you can!



## Brian Camilo Valencia Peña
# Software Developer.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2023-02-08
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2023-02-08
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/dataico/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/dataico/compare/0.1.0...0.1.1


src/invoice_spec.clj
(ns invoice-spec
  (:require
    [clojure.spec.alpha :as s]
    [clojure.data.json :as json]                            ;Add [org.clojure/data.json "0.2.6"] into the project.clj :dependencies.
    )
  (:import (java.text SimpleDateFormat)))

(use 'clojure.walk)

(s/def :customer/name string?)
(s/def :customer/email string?)
(s/def :invoice/customer (s/keys :req [:customer/name
                                       :customer/email]))

(s/def :tax/rate double?)
(s/def :tax/category #{:iva})
(s/def ::tax (s/keys :req [:tax/category
                           :tax/rate]))
(s/def :invoice-item/taxes (s/coll-of ::tax :kind vector? :min-count 1))

(s/def :invoice-item/price double?)
(s/def :invoice-item/quantity double?)
(s/def :invoice-item/sku string?)

(s/def ::invoice-item
  (s/keys :req [:invoice-item/price
                :invoice-item/quantity
                :invoice-item/sku
                :invoice-item/taxes]))

(s/def :invoice/issue-date inst?)
(s/def :invoice/items (s/coll-of ::invoice-item :kind vector? :min-count 1))

(s/def ::invoice
  (s/keys :req [:invoice/issue-date
                :invoice/customer
                :invoice/items]))

;Solution2
;This function is to change the format of the dates from"dd/MM/yyyy" to string
(def date (SimpleDateFormat. "dd/MM/yyyy"))
(defn new-date [str-date]
  (try
    (.parse date str-date)
    (catch Exception e
      (println "Error: " e))))

;This function reads the Json and evaluate if a date change is necessary.
(defn json_reader [key value]
  (case key
    :tax_category (if (= value "IVA") :iva value)           ;If it is the IVA don't change the value.
    :issue_date (new-date value)                            ;parse the date
    :tax_rate (double value)                                ;parse the date
    :payment_date (new-date value)                          ;parse the date
    value)
  )

;With the dates in the right format we can format the json file and replace the keys
(defn json-new-formater [invoice]
  (postwalk-replace
    {:issue_date   :invoice/issue-date
     :customer     :invoice/customer
     :items        :invoice/items
     :company_name :customer/name
     :email        :customer/email
     :price        :invoice-item/price
     :quantity     :invoice-item/quantity
     :sku          :invoice-item/sku
     :taxes        :invoice-item/taxes
     :tax_category :tax/category
     :tax_rate     :tax/rate} invoice))


;using json-new-formater and json_reader we call the parse in the next function.
(defn parse-json-invoice [file]
  (let [json-invoice (json/read-str (slurp file) :key-fn keyword :value-fn json_reader)]
    (get (json-new-formater json-invoice) :invoice)
    ))


;Testing and printing the parse-json-invoice function
(s/valid? ::invoice (parse-json-invoice "invoice.json"))

;PRINT (We should get True)
(print (s/valid? ::invoice (parse-json-invoice "invoice.json")))

test/dataico/core_test.clj
(ns dataico.core-test
  (:require [clojure.test :refer :all]
            [dataico.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


test/invoice_item_test.clj
(ns invoice-item_test)
(use 'clojure.test)


(defn- discount-factor [{:invoice-item/keys [discount-rate]
                         :or                {discount-rate 0}}]
  (- 1 (/ discount-rate 100.0)))

(defn subtotal
  [{:invoice-item/keys [precise-quantity precise-price discount-rate]
    :as                item
    :or                {discount-rate 0}}]
  (* precise-price precise-quantity (discount-factor item)))

;Solution3
;new function to define a test Item with its quantity, price and discount rate
(defn Item [precise-quantity precise-price discount-rate]
  {:invoice-item/precise-quantity precise-quantity
   :invoice-item/precise-price    precise-price
   :invoice-item/discount-rate    discount-rate})


;function to calculate the final price applying discounts
(defn final-price [precise-quantity precise-price discount-rate]
  (double (* precise-quantity precise-price (- 1 (/ discount-rate 100.0)))))




(deftest no-discount-price
  ;Test item definition with its values and discount.
  (let [item1 (Item 3 19.99 0)
        item2 (Item 56 15 0)
        item3 (Item 999 1.99 0)

        expected-price1 (final-price 3 19.99 0)
        expected-price2 (final-price 56 15 0)
        expected-price3 (final-price 999 1.99 0)]
    ;Apply subtotal to all the items and see if it is the same expected-price on each.
    (is (= expected-price1 (subtotal item1)))
    (is (= expected-price2 (subtotal item2)))
    (is (= expected-price3 (subtotal item3)))
    ))

(deftest discount-price
  ;Test item definition with its values and discount.
  (let [item1 (Item 3 19.99 30)
        item2 (Item 50 7.99 50)
        item3 (Item 1 99.99 99)
        expected-price1 (final-price 3 19.99 30)
        expected-price2 (final-price 50 7.99 50)
        expected-price3 (final-price 1 99.99 99)]
    ;Apply subtotal to item1, item2 and item3; and see if it is the same expected-price.
    (is (= expected-price1 (subtotal item1)))

    (is (= expected-price2 (subtotal item2)))

    (is (= expected-price3 (subtotal item3)))
    ))

(deftest multi-discount-and-price-same
  ;Test items definition with its values and discount all the same.
  (let [item1 (Item 3 19.99 30)
        item2 item1
        item3 item2
        expected-price (final-price 3 19.99 30)
        ]
    ;Apply subtotal to all items and see if the expected-prices are the same.
    (is (= expected-price (subtotal item1) (subtotal item2) (subtotal item3))
        )))

(deftest multi-discount-and-price-diff
  ;Test items definition with its values and discount all different discounts.
  (let [item1 (Item 3 19.99 30)
        item2 (Item 15 79.99 10)
        item3 (Item 8 0.99 0)
        expected-price1 (final-price 3 19.99 30)
        expected-price2 (final-price 15 79.99 10)
        expected-price3 (final-price 8 0.99 0)
        ]
    ;Apply subtotal to all items and see if the expected-prices are the same.
    (is (= expected-price1 (subtotal item1) ))
    (is (= expected-price2 (subtotal item2) ))
    (is (= expected-price3 (subtotal item3) ))
    ))

(deftest items-full-discount
  ;Test items definition with its values and discount all different discounts.
  (let [item1 (Item 35 19.99 100)
        item2 (Item 0 79.99 100)
        item3 (Item 100 0.99 100)
        expected-price1 (final-price 35 19.99 100)
        expected-price2 (final-price 0 79.99 100)
        expected-price3 (final-price 100 0.99 100)
        ]
    ;Apply subtotal to all items and see if the expected-prices are the same even with amount 0.
    (is (= expected-price1 (subtotal item1) ))
    (is (= expected-price2 (subtotal item2) ))
    (is (= expected-price3 (subtotal item3) ))
    ))


src/dataico/core.clj
(ns dataico.core
  (:gen-class)
  )


;Solution1
(def invoice (clojure.edn/read-string (slurp "src/dataico/invoice.edn"))) ;reading the invoice.edn file

;;Solution 1

;this function filters and returns the items that have a tax rate of 19
(defn tax-filter-19 [invoice]
  (->> invoice
       :invoice/items
       (filter #(some (fn [%] (= 19 (:tax/rate %))) (get-in % [:taxable/taxes])))))



;this function filters and returns the items that have a retention rate of 1
(defn ret-filter-1 [invoice]
  (->> invoice
       :invoice/items
       (filter #(some (fn [%] (= 1 (:retention/rate %))) (get-in % [:retentionable/retentions])))))


;this function returns the items that have an iva tax rate of 19 or a retention rate of 1, but not both
(defn Sym-diff-tax-ret [invoice]
  (->> invoice
       :invoice/items
       (filter (fn [item]
                 (let [filter-tax (some #(and (= 19 (:tax/rate %)) (= :iva (:tax/category %))) (get-in item [:taxable/taxes])),
                       filter-ret (some #(and (= 1 (:retention/rate %)) (= :ret_fuente (:retention/category %))) (get-in item [:retentionable/retentions]))]
                   (and (or filter-tax filter-ret) (not (and filter-tax filter-ret))))))))


;This function evaluates the other functions and print the results in clearly. The unique function was requested is Sym-diff-tax-ret. However other 2 functions are usefully to understand the sequence.
(defn evaluate
  []
  (do
    (println "Items with a tax rate of 19: " (tax-filter-19 invoice))
    (println "Items with a retention rate of 1: " (ret-filter-1 invoice))
    (println "Items with a retention rate of 1 or tax of 19 but not both: " (Sym-diff-tax-ret invoice))
    )
  )


(defn -main
  []
   (evaluate)
  )


src/invoice_item.clj
(ns invoice-item)
(use 'clojure.test)


(defn- discount-factor [{:invoice-item/keys [discount-rate]
                         :or                {discount-rate 0}}]
  (- 1 (/ discount-rate 100.0)))

(defn subtotal
  [{:invoice-item/keys [precise-quantity precise-price discount-rate]
    :as                item
    :or                {discount-rate 0}}]
  (* precise-price precise-quantity (discount-factor item)))

;Solution3
;new function to define a test Item with its quantity, price and discount rate
(defn Item [precise-quantity precise-price discount-rate]
  {:invoice-item/precise-quantity precise-quantity
   :invoice-item/precise-price    precise-price
   :invoice-item/discount-rate    discount-rate})


;function to calculate the final price applying discounts
(defn final-price [precise-quantity precise-price discount-rate]
  (double (* precise-quantity precise-price (- 1 (/ discount-rate 100.0)))))




(deftest no-discount-price
  ;Test item definition with its values and discount.
  (let [item1 (Item 3 19.99 0)
        item2 (Item 56 15 0)
        item3 (Item 999 1.99 0)

        expected-price1 (final-price 3 19.99 0)
        expected-price2 (final-price 56 15 0)
        expected-price3 (final-price 999 1.99 0)]
    ;Apply subtotal to all the items and see if it is the same expected-price on each.
        (is (= expected-price1 (subtotal item1)))
        (is (= expected-price2 (subtotal item2)))
        (is (= expected-price3 (subtotal item3)))
    ))

(deftest discount-price
  ;Test item definition with its values and discount.
  (let [item1 (Item 3 19.99 30)
        item2 (Item 50 7.99 50)
        item3 (Item 1 99.99 99)
        expected-price1 (final-price 3 19.99 30)
        expected-price2 (final-price 50 7.99 50)
        expected-price3 (final-price 1 99.99 99)]
    ;Apply subtotal to item1, item2 and item3; and see if it is the same expected-price.
    (is (= expected-price1 (subtotal item1)))

    (is (= expected-price2 (subtotal item2)))

    (is (= expected-price3 (subtotal item3)))
    ))

(deftest multi-discount-and-price-same
  ;Test items definition with its values and discount all the same.
  (let [item1 (Item 3 19.99 30)
        item2 item1
        item3 item2
        expected-price (final-price 3 19.99 30)
        ]
    ;Apply subtotal to all items and see if the expected-prices are the same.
    (is (= expected-price (subtotal item1) (subtotal item2) (subtotal item3))
        )))

(deftest multi-discount-and-price-diff
  ;Test items definition with its values and discount all different discounts.
  (let [item1 (Item 3 19.99 30)
        item2 (Item 15 79.99 10)
        item3 (Item 8 0.99 0)
        expected-price1 (final-price 3 19.99 30)
        expected-price2 (final-price 15 79.99 10)
        expected-price3 (final-price 8 0.99 0)
        ]
    ;Apply subtotal to all items and see if the expected-prices are the same.
    (is (= expected-price1 (subtotal item1) ))
    (is (= expected-price2 (subtotal item2) ))
    (is (= expected-price3 (subtotal item3) ))
    ))

(deftest items-full-discount
  ;Test items definition with its values and discount all different discounts.
  (let [item1 (Item 35 19.99 100)
        item2 (Item 0 79.99 100)
        item3 (Item 100 0.99 100)
        expected-price1 (final-price 35 19.99 100)
        expected-price2 (final-price 0 79.99 100)
        expected-price3 (final-price 100 0.99 100)
        ]
    ;Apply subtotal to all items and see if the expected-prices are the same even with amount 0.
    (is (= expected-price1 (subtotal item1) ))
    (is (= expected-price2 (subtotal item2) ))
    (is (= expected-price3 (subtotal item3) ))
    ))
